File: src/retrieval/engine.ts
/**
 * Retrieval Engine (v3 — 5-Signal Fusion)
 *
 * Implements hybrid retrieval with 5 parallel signals:
 *   1. Semantic vector search (pgvector ANN)
 *   2. Keyword BM25 search (Postgres tsvector)
 *   3. Entity matching (boost results sharing entities with query)
 *   4. Temporal scoring (time-aware relevance, valid/superseded)
 *   5. Graph expansion (AGE relationship traversal, conditional)
 *
 * Also searches the Fact layer (atomic facts) for quick answers.
 * Returns a mix of facts (instant answers) and chunks (full context).
 *
 * Target: <180ms p99 (warm), <5ms (cached)
 */

import { v4 as uuidv4 } from 'uuid';
import { createChildLogger } from '../utils/logger.js';
import { EmbeddingClient } from '../utils/embedding.js';
import { ChunkRepository } from '../storage/chunk-repository.js';
import { FactRepository } from '../storage/fact-repository.js';
import { GraphRepository } from '../storage/graph-repository.js';
import { SearchIndex } from '../storage/search-index.js';
import { SearchCache } from '../storage/cache.js';
import { RankingEngine } from './ranking.js';
import { PermissionFilter } from './permission-filter.js';
import {
  type SearchRequest,
  type SearchResponse,
  type SearchResultItem,
  type Chunk,
  type MemoryFact,
} from '../models/index.js';

const logger = createChildLogger({ module: 'retrieval-engine' });

// ─── Internal Candidate Type (v3: supports both facts and chunks) ────────────

interface RetrievalCandidate {
  chunk: Chunk;
  facts?: MemoryFact[];
  scores: {
    semantic: number;
    keyword: number;
    entityMatch: number;
    temporal: number;
    graphRelevance: number;
  };
  source: 'vector' | 'keyword' | 'entity' | 'temporal' | 'graph';
}

// ─── Query Entity Extraction ─────────────────────────────────────────────────

function extractQueryEntities(query: string): string[] {
  const entities: string[] = [];
  const seen = new Set<string>();
  const techMatches = query.match(
    /\b(AWS|S3|EC2|Lambda|DynamoDB|Kafka|Redis|Postgres(?:QL)?|MongoDB|Docker|Kubernetes|K8s|React|Node\.?js?|TypeScript|Python|Go|Rust|GraphQL|REST|gRPC|Terraform|CDK|IAM|VPC|ECS|EKS|RDS|SQS|SNS|CloudFront|CloudWatch|Next\.?js|Express|Fastify)\b/gi
  );
  if (techMatches) {
    for (const t of techMatches) {
      if (!seen.has(t.toLowerCase())) { entities.push(t); seen.add(t.toLowerCase()); }
    }
  }
  const quoted = query.match(/"([^"]+)"/g);
  if (quoted) {
    for (const q of quoted) {
      const term = q.replace(/"/g, '');
      if (!seen.has(term.toLowerCase())) { entities.push(term); seen.add(term.toLowerCase()); }
    }
  }
  return entities;
}

// ─── Retrieval Engine ────────────────────────────────────────────────────────

export class RetrievalEngine {
  private embeddingClient: EmbeddingClient;
  private chunkRepo: ChunkRepository;
  private factRepo: FactRepository;
  private graphRepo: GraphRepository;
  private searchIndex: SearchIndex;
  private searchCache: SearchCache;
  private rankingEngine: RankingEngine;
  private permissionFilter: PermissionFilter;

  constructor() {
    this.embeddingClient = new EmbeddingClient();
    this.chunkRepo = new ChunkRepository();
    this.factRepo = new FactRepository();
    this.graphRepo = new GraphRepository();
    this.searchIndex = new SearchIndex();
    this.searchCache = new SearchCache();
    this.rankingEngine = new RankingEngine();
    this.permissionFilter = new PermissionFilter();
  }

  /**
   * Main search entry point. Orchestrates the full hybrid retrieval pipeline.
   */
  async search(request: SearchRequest): Promise<SearchResponse> {
    const searchId = uuidv4();
    const startTime = Date.now();

    logger.info({
      searchId,
      query: request.query.substring(0, 100),
      strategy: request.strategy,
      topK: request.topK,
    }, 'Search started');

    // 1. Check cache first
    const cached = await this.searchCache.get(request);
    if (cached) {
      return { ...cached, cached: true, latencyMs: Date.now() - startTime };
    }

    // 2. Pre-process: entities for signal 3, and the query vector.
    // Embedding runs before the database scope opens: it may call a remote
    // provider, and holding a pooled connection across that wait would starve
    // the pool under load.
    const queryEntities = extractQueryEntities(request.query);
    const needsVector = request.strategy !== 'keyword';
    const queryEmbedding = needsVector
      ? await this.embeddingClient.embed(request.query)
      : null;

    // 3. Execute the strategy. Signals run concurrently on separate pooled
    // connections: measured on a low-latency link, overlapping the queries beats
    // sharing one connection to save the per-statement RLS setup round trips.
    let candidates: RetrievalCandidate[];
    switch (request.strategy) {
      case 'semantic':
        candidates = await this.semanticSearch(request, queryEmbedding!);
        break;
      case 'keyword':
        candidates = await this.keywordSearch(request);
        break;
      case 'graph':
        candidates = await this.graphSearch(request, queryEmbedding!);
        break;
      case 'hybrid':
      default:
        candidates = await this.hybridSearch(request, queryEntities, queryEmbedding!);
        break;
    }

    // 4. Permission filtering
    candidates = await this.permissionFilter.filter(candidates, {
      userId: request.developerId,
      organizationId: request.organizationId,
      teamIds: request.teamIds,
      roles: request.roles,
      repositoryAccess: request.repositoryAccess,
    });

    // 4. Rank and score
    const ranked = await this.rankingEngine.rank(candidates, request);

    // 5. Take top-K results
    const topResults = ranked.slice(request.offset, request.offset + request.topK);

    // 6. Format response
    const results: SearchResultItem[] = topResults.map(candidate => this.formatResult(candidate));

    // 7. Estimate token usage
    const estimatedTokens = results.reduce((sum, r) => {
      return sum + Math.ceil((r.content?.length ?? r.summary.length) / 4);
    }, 0);

    const response: SearchResponse = {
      results,
      totalCount: ranked.length,
      query: request.query,
      strategy: request.strategy,
      latencyMs: Date.now() - startTime,
      cached: false,
      estimatedTokens,
      relatedQueries: [], // TODO: generate related queries
    };

    // 8. Cache the response
    await this.searchCache.set(request, response);
    await this.searchCache.trackQuery(request.query, request.organizationId);

    logger.info({
      searchId,
      resultCount: results.length,
      latencyMs: response.latencyMs,
      estimatedTokens,
    }, 'Search completed');

    return response;
  }

  /**
   * Find chunks similar to a given chunk (for "related knowledge" features).
   */
  async findSimilar(chunkId: string, limit: number = 5): Promise<SearchResultItem[]> {
    const chunk = await this.chunkRepo.findById(chunkId);
    if (!chunk || !chunk.embedding) return [];

    const similar = await this.chunkRepo.searchByVector(chunk.embedding, {
      limit: limit + 1, // +1 because we'll exclude self
      organizationId: chunk.organizationId,
    });

    return similar
      .filter(s => s.id !== chunkId)
      .slice(0, limit)
      .map(s => this.formatResult({
        chunk: s,
        scores: { semantic: s.similarity, keyword: 0, entityMatch: 0, temporal: 0, graphRelevance: 0 },
        source: 'vector',
      }));
  }

  // ─── Search Strategies ─────────────────────────────────────────────────────

  /**
   * Pure semantic vector search.
   * Fast, good for conceptual similarity, may miss exact terms.
   */
  private async semanticSearch(
    request: SearchRequest,
    queryEmbedding: number[],
  ): Promise<RetrievalCandidate[]> {
    // Vector ANN search
    const results = await this.chunkRepo.searchByVector(queryEmbedding, {
      limit: 50,
      organizationId: request.organizationId,
      repositoryFilter: request.filters?.repositories,
      languageFilter: request.filters?.languages,
      minQualityScore: request.filters?.minQualityScore,
    });

    return results.map(r => ({
      chunk: r,
      scores: { semantic: r.similarity, keyword: 0, entityMatch: 0, temporal: 0, graphRelevance: 0 },
      source: 'vector' as const,
    }));
  }

  /**
   * Pure keyword BM25 search.
   * Good for exact terms, error messages, function names.
   */
  private async keywordSearch(request: SearchRequest): Promise<RetrievalCandidate[]> {
    const results = await this.searchIndex.search({
      query: request.query,
      organizationId: request.organizationId,
      filters: {
        repositories: request.filters?.repositories,
        languages: request.filters?.languages,
        types: request.filters?.types,
        minQualityScore: request.filters?.minQualityScore,
      },
      limit: 50,
    });

    // Load matched chunks in one round trip, preserving relevance order.
    const chunks = await this.chunkRepo.findByIds(
      results.map(hit => hit.id),
      request.organizationId,
    );

    const topScore = results[0]?.score ?? 1;
    const candidates: RetrievalCandidate[] = [];
    for (const hit of results) {
      const chunk = chunks.get(hit.id);
      if (!chunk) continue;
      candidates.push({
        chunk,
        scores: {
          semantic: 0,
          keyword: hit.score / topScore,
          entityMatch: 0,
          temporal: 0,
          graphRelevance: 0,
        },
        source: 'keyword',
      });
    }

    return candidates;
  }

  /**
   * Graph-based search.
   * Starts from query entities and traverses relationships.
   */
  private async graphSearch(
    request: SearchRequest,
    queryEmbedding: number[],
  ): Promise<RetrievalCandidate[]> {
    // Extract entities from query for graph seed nodes
    const seedNodeIds: string[] = [];

    // Use repository as seed if provided
    if (request.context?.repository) {
      seedNodeIds.push(request.context.repository);
    }

    // Use language/framework as seed
    if (request.context?.language) {
      seedNodeIds.push(request.context.language.toLowerCase());
    }
    if (request.context?.frameworks) {
      for (const fw of request.context.frameworks) {
        seedNodeIds.push(fw.toLowerCase());
      }
    }

    if (seedNodeIds.length === 0) {
      // Fall back to semantic search if no graph seeds
      return this.semanticSearch(request, queryEmbedding);
    }

    const expansion = await this.graphRepo.expand({
      seedNodeIds,
      organizationId: request.organizationId,
      maxDepth: 2,
      targetNodeTypes: ['chunk', 'knowledge'],
      limit: 30,
      minWeight: 0.3,
    });

    // Load graph-discovered chunks in one round trip, preserving expansion order.
    const chunks = await this.chunkRepo.findByIds(
      expansion.relatedContentIds,
      request.organizationId,
    );

    const candidates: RetrievalCandidate[] = [];
    for (const contentId of expansion.relatedContentIds) {
      const chunk = chunks.get(contentId);
      if (!chunk) continue;
      candidates.push({
        chunk,
        scores: { semantic: 0, keyword: 0, entityMatch: 0, temporal: 0, graphRelevance: 0.7 },
        source: 'graph',
      });
    }

    return candidates;
  }

  /**
   * Hybrid search — 5-signal fusion (v3).
   * Runs Semantic + Keyword + Entity in parallel, applies Temporal to all,
   * conditionally expands Graph when semantic confidence is low.
   */
  private async hybridSearch(
    request: SearchRequest,
    queryEntities: string[],
    queryEmbedding: number[],
  ): Promise<RetrievalCandidate[]> {
    // Three signals in parallel, each on its own pooled connection.
    const [semanticResults, keywordResults, entityResults] = await Promise.all([
      this.semanticSearch(request, queryEmbedding),
      this.keywordSearch(request),
      this.entitySearch(queryEntities, request.organizationId),
    ]);

    // Reciprocal Rank Fusion (RRF) to combine results
    const fusedMap = new Map<string, RetrievalCandidate>();

    for (const candidate of semanticResults) {
      fusedMap.set(candidate.chunk.id, candidate);
    }

    for (const candidate of keywordResults) {
      const existing = fusedMap.get(candidate.chunk.id);
      if (existing) {
        existing.scores.keyword = candidate.scores.keyword;
      } else {
        fusedMap.set(candidate.chunk.id, candidate);
      }
    }

    for (const candidate of entityResults) {
      const existing = fusedMap.get(candidate.chunk.id);
      if (existing) {
        existing.scores.entityMatch = candidate.scores.entityMatch;
        if (candidate.facts) existing.facts = candidate.facts;
      } else {
        fusedMap.set(candidate.chunk.id, candidate);
      }
    }

    // Signal 4: Temporal — apply to ALL candidates (free, metadata-based)
    for (const candidate of fusedMap.values()) {
      candidate.scores.temporal = this.computeTemporalScore(candidate.chunk);
    }

    // Signal 5: Graph — ADAPTIVE, only when semantic confidence is low
    const topSemanticScore = semanticResults[0]?.scores.semantic ?? 0;
    const shouldExpandGraph = topSemanticScore < 0.85 && request.context?.repository;

    if (shouldExpandGraph) {
      const graphResults = await this.graphSearch(request, queryEmbedding);
      for (const candidate of graphResults) {
        const existing = fusedMap.get(candidate.chunk.id);
        if (existing) {
          existing.scores.graphRelevance = candidate.scores.graphRelevance;
        } else {
          candidate.scores.temporal = this.computeTemporalScore(candidate.chunk);
          fusedMap.set(candidate.chunk.id, candidate);
        }
      }
      logger.debug({ topSemanticScore, graphResultCount: graphResults.length }, 'Graph expansion triggered');
    }

    return [...fusedMap.values()];
  }

  /**
   * Signal 3: Entity matching.
   * Searches the fact layer for entities, then loads parent chunks.
   */
  private async entitySearch(queryEntities: string[], organizationId: string): Promise<RetrievalCandidate[]> {
    if (queryEntities.length === 0) return [];

    const matchingFacts = await this.factRepo.findByEntities(queryEntities, organizationId, {
      limit: 20,
      onlyValid: true,
    });

    // Resolve every source chunk in one round trip before scoring.
    const sourceChunkIds = matchingFacts
      .map(fact => fact.sourceChunkId)
      .filter((id): id is string => typeof id === 'string');
    const chunks = await this.chunkRepo.findByIds(sourceChunkIds, organizationId);

    const candidates: RetrievalCandidate[] = [];
    const seenChunks = new Set<string>();
    const queryLower = queryEntities.map(entity => entity.toLowerCase());

    for (const fact of matchingFacts) {
      if (!fact.sourceChunkId || seenChunks.has(fact.sourceChunkId)) continue;
      const chunk = chunks.get(fact.sourceChunkId);
      if (!chunk) continue;
      seenChunks.add(fact.sourceChunkId);

      const chunkEntityNames = chunk.entities.map(e => e.name.toLowerCase());
      const overlap = queryLower.filter(entity => chunkEntityNames.includes(entity)).length;
      const unionSize = new Set([...queryLower, ...chunkEntityNames]).size;
      const entityScore = unionSize > 0 ? Math.min((overlap / unionSize) * 1.5, 1.0) : 0;

      candidates.push({
        chunk,
        facts: [fact],
        scores: { semantic: 0, keyword: 0, entityMatch: entityScore, temporal: 0, graphRelevance: 0 },
        source: 'entity',
      });
    }

    return candidates;
  }

  /**
   * Signal 4: Temporal scoring (Pieces-inspired).
   * Recent + recently-accessed = higher. Stale/archived = lower.
   */
  private computeTemporalScore(chunk: Chunk): number {
    const ageMs = Date.now() - new Date(chunk.createdAt).getTime();
    const ageDays = ageMs / (1000 * 60 * 60 * 24);
    let score = Math.exp(-ageDays / 130); // 90-day half-life

    if (chunk.lastAccessedAt) {
      const lastAccessDays = (Date.now() - new Date(chunk.lastAccessedAt).getTime()) / 86400000;
      if (lastAccessDays < 7) score = Math.min(score + 0.2, 1.0);
    }

    if (chunk.confidence === 'low') score *= 0.6;
    if (chunk.confidence === 'archived') score *= 0.3;
    return score;
  }

  // ─── Result Formatting ─────────────────────────────────────────────────────

  private formatResult(candidate: RetrievalCandidate & { finalScore?: number }): SearchResultItem {
    const chunk = candidate.chunk;

    return {
      id: chunk.id,
      type: 'chunk',
      title: chunk.title,
      summary: chunk.summary,
      content: chunk.content,
      finalScore: candidate.finalScore ?? Math.max(
        candidate.scores.semantic,
        candidate.scores.keyword,
        candidate.scores.graphRelevance,
      ),
      scores: {
        semantic: candidate.scores.semantic,
        keyword: candidate.scores.keyword,
        freshness: candidate.scores.temporal,
        repositoryMatch: 0,
        authorReputation: 0,
        usageCount: Math.min(chunk.usageCount / 100, 1),
        qualityScore: chunk.qualityScore,
      },
      repository: chunk.repository,
      language: chunk.language,
      frameworks: chunk.frameworks,
      author: { id: chunk.authorId },
      citations: [{
        type: 'conversation',
        reference: `Session ${chunk.sessionId}`,
      }],
      codeSnippets: chunk.codeReferences.slice(0, 3).map(ref => ({
        language: ref.language,
        code: ref.snippet,
        filePath: ref.filePath,
      })),
      createdAt: chunk.createdAt,
      lastUsedAt: chunk.lastAccessedAt,
    };
  }
}

File: go/internal/retrieval/engine.go
// Package retrieval implements the 5-signal hybrid search pipeline.
//
// Architecture:
//  1. Query → embed (or cache hit)
//  2. 3-5 parallel DB signals (vector, keyword, entity, temporal, graph)
//  3. RRF fusion → composite ranking → diversity → token budget packing
//  4. Return results
//
// The ranking math runs inline in Go (fast enough for <200 candidates).
// For batch operations (1000+ candidates), we can optionally call the Rust
// native module via subprocess.
package retrieval

import (
	"context"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"log/slog"
	"math"
	"sort"
	"sync"
	"time"

	"github.com/mihaibalaci/synapse/internal/auth"
	"github.com/mihaibalaci/synapse/internal/models"
	"github.com/mihaibalaci/synapse/internal/storage"
)

// Embedder turns a query into a vector so the semantic signal can run. It is
// an interface rather than a concrete type to keep this package independent of
// the ingestion package and its provider configuration.
type Embedder interface {
	Embed(ctx context.Context, text string) ([]float64, error)
}

// Engine orchestrates the full retrieval pipeline.
type Engine struct {
	db    *storage.DB
	cache *storage.Cache

	chunks *storage.ChunkRepo
	facts  *storage.FactRepo

	// embedder may be nil, in which case the semantic signal is skipped and
	// retrieval falls back to keyword and entity matching.
	embedder Embedder
}

// NewEngine creates a retrieval engine with the given dependencies. Passing a
// nil embedder disables the semantic signal.
func NewEngine(db *storage.DB, cache *storage.Cache, chunks *storage.ChunkRepo, facts *storage.FactRepo, embedder Embedder) *Engine {
	return &Engine{db: db, cache: cache, chunks: chunks, facts: facts, embedder: embedder}
}

// Search executes the full hybrid retrieval pipeline.
func (e *Engine) Search(ctx context.Context, req *models.SearchRequest, claims *auth.Claims) (*models.SearchResponse, error) {
	start := time.Now()

	// 1. Check cache
	cacheKey := e.buildCacheKey(req, claims)
	// Cache failures degrade to a normal search; malformed cached JSON is
	// deleted so every request does not repeatedly pay the decode failure.
	if cached, err := e.cache.GetCached(ctx, cacheKey); err != nil {
		slog.Warn("Search cache read failed", "error", err)
	} else if cached != nil {
		var resp models.SearchResponse
		if err := json.Unmarshal(cached, &resp); err == nil {
			resp.Cached = true
			resp.LatencyMs = time.Since(start).Milliseconds()
			return &resp, nil
		}
		slog.Warn("Discarding malformed cached search response", "key", cacheKey)
		_ = e.cache.Client.Del(ctx, cacheKey).Err()
	}

	orgID := claims.OrganizationID
	topK := req.TopK
	if topK == 0 {
		topK = 5
	}

	// 2. Run signals in parallel
	type signalResult struct {
		candidates []candidate
		err        error
	}

	var wg sync.WaitGroup
	semanticCh := make(chan signalResult, 1)
	keywordCh := make(chan signalResult, 1)
	entityCh := make(chan signalResult, 1)
	graphCh := make(chan signalResult, 1)

	// Signal 1: Semantic (vector ANN over pgvector).
	wg.Add(1)
	go func() {
		defer wg.Done()

		// Without an embedder there is no query vector, and passing nil would
		// make the similarity NULL for every row. Skip the signal instead of
		// polluting the fusion with meaningless scores.
		if e.embedder == nil {
			semanticCh <- signalResult{}
			return
		}

		queryVec, err := e.embedder.Embed(ctx, req.Query)
		if err != nil {
			slog.Warn("Query embedding failed; semantic signal skipped", "error", err)
			semanticCh <- signalResult{}
			return
		}

		results, err := e.chunks.SearchByVector(ctx, queryVec, orgID, 50)
		candidates := make([]candidate, len(results))
		for i, r := range results {
			candidates[i] = candidate{
				ID: r.ID, Title: r.Title, Summary: r.Summary, Content: r.Content,
				TokenCount: r.TokenCount, QualityScore: r.QualityScore,
				UsageCount: r.UsageCount, Confidence: r.Confidence,
				Repository: r.Repository, CreatedAt: r.CreatedAt,
				Scores: scores{Semantic: r.Similarity},
			}
		}
		semanticCh <- signalResult{candidates: candidates, err: err}
	}()

	// Signal 2: Keyword (BM25)
	wg.Add(1)
	go func() {
		defer wg.Done()
		results, err := e.chunks.SearchByKeyword(ctx, req.Query, orgID, 50)
		candidates := make([]candidate, len(results))
		for i, r := range results {
			candidates[i] = candidate{
				ID: r.ID, Title: r.Title, Summary: r.Summary, Content: r.Content,
				TokenCount: r.TokenCount, QualityScore: r.QualityScore,
				UsageCount: r.UsageCount, Confidence: r.Confidence,
				Repository: r.Repository, CreatedAt: r.CreatedAt,
				Scores: scores{Keyword: r.Similarity},
			}
		}
		keywordCh <- signalResult{candidates: candidates, err: err}
	}()

	// Signal 3: Entity match (facts)
	wg.Add(1)
	go func() {
		defer wg.Done()
		entities := extractQueryEntities(req.Query)
		if len(entities) == 0 {
			entityCh <- signalResult{}
			return
		}
		facts, err := e.facts.FindByEntities(ctx, entities, orgID, 20)
		candidates := make([]candidate, 0, len(facts))
		for _, f := range facts {
			candidates = append(candidates, candidate{
				ID: f.ID, Title: f.Content, Summary: f.Content,
				Scores: scores{EntityMatch: 0.7},
			})
		}
		entityCh <- signalResult{candidates: candidates, err: err}
	}()

	// Signal 4: Graph neighbors — boost chunks whose entities share graph edges
	// with the query entities.
	wg.Add(1)
	go func() {
		defer wg.Done()
		entities := extractQueryEntities(req.Query)
		if len(entities) == 0 {
			graphCh <- signalResult{}
			return
		}
		rows, err := e.db.Query(ctx, `
			SELECT DISTINCT c.id, c.title, c.summary, c.content, c.token_count,
				c.quality_score, c.usage_count, c.confidence, COALESCE(c.repository,''), c.created_at,
				ge.weight / 10.0 AS score
			FROM graph_nodes gn
			JOIN graph_edges ge ON ge.source_id = gn.id OR ge.target_id = gn.id
			JOIN graph_nodes neighbor ON neighbor.id = CASE WHEN ge.source_id = gn.id THEN ge.target_id ELSE ge.source_id END
			JOIN memory_facts mf ON neighbor.name = ANY(mf.entities) AND mf.organization_id = $2
			JOIN chunks c ON c.id = mf.source_chunk_id AND c.confidence <> 'archived' AND c.searchable_status = 'searchable'
			WHERE gn.organization_id = $2 AND gn.name = ANY($1::text[])
			ORDER BY ge.weight DESC
			LIMIT 20`, entities, orgID)
		if err != nil {
			graphCh <- signalResult{err: err}
			return
		}
		defer rows.Close()
		var candidates []candidate
		for rows.Next() {
			var c candidate
			if err := rows.Scan(&c.ID, &c.Title, &c.Summary, &c.Content, &c.TokenCount,
				&c.QualityScore, &c.UsageCount, &c.Confidence, &c.Repository, &c.CreatedAt, &c.Scores.GraphRelevance); err != nil {
				continue
			}
			candidates = append(candidates, c)
		}
		graphCh <- signalResult{candidates: candidates}
	}()

	wg.Wait()

	// 3. Collect results
	sem := <-semanticCh
	kw := <-keywordCh
	ent := <-entityCh
	graph := <-graphCh

	if sem.err != nil {
		slog.Warn("Semantic search failed", "error", sem.err)
	}
	if kw.err != nil {
		slog.Warn("Keyword search failed", "error", kw.err)
	}
	if graph.err != nil {
		slog.Warn("Graph search failed", "error", graph.err)
	}

	// 4. RRF Fusion
	fused := rrfFuse(sem.candidates, kw.candidates, ent.candidates, graph.candidates)

	// 5. Composite ranking
	queryRepo := ""
	if req.Context != nil {
		queryRepo = req.Context.Repository
	}
	ranked := rankCandidates(fused, queryRepo)

	// 6. Apply diversity
	ranked = applyDiversity(ranked)

	// 7. Token budget packing or top-K
	var selected []candidate
	if req.MaxTokens > 0 {
		selected = packByBudget(ranked, req.MaxTokens)
	} else {
		end := topK
		if end > len(ranked) {
			end = len(ranked)
		}
		selected = ranked[:end]
	}

	// 8. Format response
	results := make([]models.SearchResult, len(selected))
	estimatedTokens := 0
	for i, c := range selected {
		content := ""
		if req.IncludeContent {
			content = c.Content
		}
		results[i] = models.SearchResult{
			ID:         c.ID,
			Type:       "chunk",
			Title:      c.Title,
			Summary:    c.Summary,
			Content:    content,
			FinalScore: c.FinalScore,
			Repository: c.Repository,
			CreatedAt:  c.CreatedAt.Format(time.RFC3339),
		}
		estimatedTokens += c.TokenCount
	}

	resp := &models.SearchResponse{
		Results:         results,
		TotalCount:      len(fused),
		Query:           req.Query,
		Strategy:        req.Strategy,
		LatencyMs:       time.Since(start).Milliseconds(),
		Cached:          false,
		EstimatedTokens: estimatedTokens,
		Observations:    []models.Observation{},
	}

	// Cache the response
	if data, err := json.Marshal(resp); err == nil {
		if err := e.cache.SetCached(ctx, cacheKey, data, 5*time.Minute); err != nil {
			slog.Warn("Search cache write failed", "error", err)
		}
	}
	e.cache.IncrementPopular(ctx, orgID, req.Query)

	slog.Info("Search completed",
		"query", truncate(req.Query, 50),
		"results", len(results),
		"latencyMs", resp.LatencyMs,
	)

	return resp, nil
}

// ─── Internal Types ──────────────────────────────────────────────────────────

type scores struct {
	Semantic       float64
	Keyword        float64
	EntityMatch    float64
	Temporal       float64
	GraphRelevance float64
}

type candidate struct {
	ID           string
	Title        string
	Summary      string
	Content      string
	TokenCount   int
	QualityScore float64
	UsageCount   int
	Confidence   string
	Repository   string
	SessionID    string
	CreatedAt    time.Time
	Scores       scores
	FinalScore   float64
}

// ─── RRF Fusion ──────────────────────────────────────────────────────────────

func rrfFuse(lists ...[]candidate) []candidate {
	const k = 60.0
	scoreMap := make(map[string]*candidate)
	rrfScores := make(map[string]float64)

	for _, list := range lists {
		for rank, c := range list {
			rrfScores[c.ID] += 1.0 / (k + float64(rank) + 1.0)
			if _, exists := scoreMap[c.ID]; !exists {
				copy := c
				scoreMap[c.ID] = &copy
			} else {
				// Merge scores
				existing := scoreMap[c.ID]
				existing.Scores.Semantic = math.Max(existing.Scores.Semantic, c.Scores.Semantic)
				existing.Scores.Keyword = math.Max(existing.Scores.Keyword, c.Scores.Keyword)
				existing.Scores.EntityMatch = math.Max(existing.Scores.EntityMatch, c.Scores.EntityMatch)
			}
		}
	}

	result := make([]candidate, 0, len(scoreMap))
	for id, c := range scoreMap {
		c.FinalScore = rrfScores[id]
		result = append(result, *c)
	}

	sort.Slice(result, func(i, j int) bool {
		return result[i].FinalScore > result[j].FinalScore
	})
	return result
}

// ─── Composite Ranking ───────────────────────────────────────────────────────

func rankCandidates(candidates []candidate, queryRepo string) []candidate {
	now := time.Now()

	for i := range candidates {
		c := &candidates[i]
		s := c.Scores

		// Signal weights (sum to ~1.0 before confidence multiplier)
		semantic := s.Semantic * 0.25
		keyword := s.Keyword * 0.10
		entity := s.EntityMatch * 0.08
		graph := s.GraphRelevance * 0.07

		// Temporal decay: 30-day half-life for recent relevance, floor at 0.1
		ageDays := now.Sub(c.CreatedAt).Hours() / 24.0
		temporalDecay := math.Max(math.Exp(-0.693*ageDays/30.0), 0.1)
		freshness := temporalDecay * 0.15

		// Repository match boost
		repoMatch := 0.0
		if queryRepo != "" && c.Repository == queryRepo {
			repoMatch = 0.12
		}

		// Usage with decay: high usage is good but old unused content decays
		usageRaw := math.Min(float64(c.UsageCount)/50.0, 1.0)
		usageDecayed := usageRaw * temporalDecay
		usage := usageDecayed * 0.10

		// Quality score
		quality := c.QualityScore * 0.08

		// Importance score: combines multiple signals into a compound relevance
		importance := (semantic + keyword + entity + graph + freshness + repoMatch + usage + quality)

		// Confidence multiplier
		confMult := 1.0
		switch c.Confidence {
		case "low":
			confMult = 0.6
		case "archived":
			confMult = 0.2
		case "high":
			confMult = 1.1
		}

		c.FinalScore = importance * confMult
		c.Scores.Temporal = temporalDecay
	}

	sort.Slice(candidates, func(i, j int) bool {
		return candidates[i].FinalScore > candidates[j].FinalScore
	})
	return candidates
}

// ─── Diversity ───────────────────────────────────────────────────────────────

func applyDiversity(candidates []candidate) []candidate {
	sessionCounts := make(map[string]int)
	for i := range candidates {
		sid := candidates[i].SessionID
		if sid == "" {
			continue
		}
		count := sessionCounts[sid]
		if count >= 2 {
			candidates[i].FinalScore *= 0.5
		}
		sessionCounts[sid] = count + 1
	}
	sort.Slice(candidates, func(i, j int) bool {
		return candidates[i].FinalScore > candidates[j].FinalScore
	})
	return candidates
}

// ─── Token Budget Packing ────────────────────────────────────────────────────

func packByBudget(candidates []candidate, maxTokens int) []candidate {
	var result []candidate
	remaining := maxTokens

	for _, c := range candidates {
		tokens := c.TokenCount
		if tokens == 0 {
			tokens = len(c.Content) / 4
		}
		if tokens > remaining {
			if len(result) == 0 {
				result = append(result, c)
			}
			break
		}
		remaining -= tokens
		result = append(result, c)
	}
	return result
}

// ─── Helpers ─────────────────────────────────────────────────────────────────

func (e *Engine) buildCacheKey(req *models.SearchRequest, claims *auth.Claims) string {
	// Include every request field that can alter the response plus the complete
	// authorization context. Omitting maxTokens/includeContent previously let a
	// summary-only response satisfy a later full-content request.
	data, _ := json.Marshal(map[string]any{
		"request": req,
		"auth": map[string]any{
			"organizationId":   claims.OrganizationID,
			"userId":           claims.UserID,
			"teamIds":          claims.TeamIDs,
			"roles":            claims.Roles,
			"repositoryAccess": claims.RepositoryAccess,
		},
	})
	hash := sha256.Sum256(data)
	return "search:" + hex.EncodeToString(hash[:16])
}

func extractQueryEntities(query string) []string {
	// Simple heuristic: extract capitalized tech terms
	// In production, use the Rust batchEntityOverlap for precision
	var entities []string
	words := splitWords(query)
	techTerms := map[string]bool{
		"aws": true, "s3": true, "lambda": true, "kafka": true, "redis": true,
		"postgres": true, "docker": true, "kubernetes": true, "react": true,
		"typescript": true, "python": true, "go": true, "rust": true, "graphql": true,
		"terraform": true, "dynamodb": true, "ec2": true, "ecs": true, "vpc": true,
	}
	for _, w := range words {
		lower := toLower(w)
		if techTerms[lower] {
			entities = append(entities, w)
		}
	}
	return entities
}

func splitWords(s string) []string {
	var words []string
	current := ""
	for _, c := range s {
		if c == ' ' || c == '\t' || c == '\n' {
			if current != "" {
				words = append(words, current)
				current = ""
			}
		} else {
			current += string(c)
		}
	}
	if current != "" {
		words = append(words, current)
	}
	return words
}

func toLower(s string) string {
	result := make([]byte, len(s))
	for i := 0; i < len(s); i++ {
		c := s[i]
		if c >= 'A' && c <= 'Z' {
			c += 32
		}
		result[i] = c
	}
	return string(result)
}

func truncate(s string, maxLen int) string {
	if len(s) <= maxLen {
		return s
	}
	return s[:maxLen] + "..."
}
